File: ios/MetanavCore/Sources/MetanavCore/ObstacleTracker.swift
import Foundation

/// A blob followed across frames so that one noisy frame can neither trigger nor cancel an alert.
public final class Track {
    public let id: Int
    public private(set) var left: Float
    public private(set) var right: Float
    public private(set) var distanceMeters: Float
    public private(set) var lastSeenFrame: Int
    public let firstSeenFrame: Int
    private var labelVotes: [String: Int] = [:]
    /// Frame indices in which this track was observed (bounded by the confirmation window).
    private var hitFrames: [Int] = []

    // Announcement bookkeeping, owned by the policy.
    var announcedCount = 0
    var lastAnnouncedDistance: Float = .infinity
    var lastAnnouncedAtMs: Int64 = AdvisoryPolicy.never
    var lastAnnouncedUrgency: Urgency = .none
    var reminders = 0

    init(id: Int, blob: Blob, frameIndex: Int) {
        self.id = id
        left = blob.left
        right = blob.right
        distanceMeters = blob.distanceMeters
        lastSeenFrame = frameIndex
        firstSeenFrame = frameIndex
        hitFrames = [frameIndex]
    }

    public var centerX: Float { (left + right) / 2 }

    public var label: String {
        labelVotes.max { $0.value < $1.value }?.key ?? Labels.generic
    }

    func update(blob: Blob, frameIndex: Int, window: Int) {
        left = 0.5 * left + 0.5 * blob.left
        right = 0.5 * right + 0.5 * blob.right
        distanceMeters = 0.6 * distanceMeters + 0.4 * blob.distanceMeters
        lastSeenFrame = frameIndex
        hitFrames.append(frameIndex)
        hitFrames.removeAll { $0 <= frameIndex - window }
    }

    func vote(_ label: String) { labelVotes[label, default: 0] += 1 }

    func fuseDistance(_ meters: Float) {
        // Size-based estimates are coarse; blend gently and only when they roughly agree.
        let ratio = meters / distanceMeters
        if ratio >= 0.5 && ratio <= 2.0 { distanceMeters = 0.75 * distanceMeters + 0.25 * meters }
    }

    public func hitsInWindow(frameIndex: Int, window: Int) -> Int {
        hitFrames.filter { $0 > frameIndex - window }.count
    }

    public func isConfirmed(frameIndex: Int, config: ReasonerConfig) -> Bool {
        hitsInWindow(frameIndex: frameIndex, window: config.confirmWindow) >= config.confirmHits
    }

    public func overlapsCorridor(_ config: ReasonerConfig) -> Bool {
        let overlap = min(right, config.corridorRight) - max(left, config.corridorLeft)
        if overlap <= 0 { return false }
        let needed = min(0.4 * (right - left), 0.08)
        return overlap >= needed
    }
}

public final class ObstacleTracker {
    private let config: ReasonerConfig
    private var tracks: [Track] = []
    private var nextId = 1
    private var frameIndex = 0

    public init(config: ReasonerConfig) { self.config = config }

    public var active: [Track] { tracks }
    public func currentFrame() -> Int { frameIndex }

    @discardableResult
    public func update(blobs: [Blob]) -> [Track] {
        frameIndex += 1
        var unmatched = blobs
        // Greedy association by horizontal overlap, closest-first.
        for track in tracks.sorted(by: { $0.distanceMeters < $1.distanceMeters }) {
            var bestIndex: Int? = nil
            var bestScore: Float = 0
            for (i, blob) in unmatched.enumerated() {
                let score = Self.overlap1d(track.left, track.right, blob.left, blob.right)
                let centerGap = abs(track.centerX - blob.centerX)
                let s: Float = (score > 0.3 || centerGap < 0.15) ? max(score, 0.3 - centerGap) : 0
                if s > bestScore { bestScore = s; bestIndex = i }
            }
            if let i = bestIndex {
                track.update(blob: unmatched[i], frameIndex: frameIndex, window: config.confirmWindow)
                unmatched.remove(at: i)
            }
        }
        for blob in unmatched {
            tracks.append(Track(id: nextId, blob: blob, frameIndex: frameIndex))
            nextId += 1
        }
        tracks.removeAll { frameIndex - $0.lastSeenFrame > config.maxMissedFrames }
        return tracks
    }

    /// Attach detector labels (and size-based distances) to tracks seen this frame.
    public func applyDetections(_ detections: [Detection], ground: GroundPlaneModel, frameHeightPx: Int) {
        guard !detections.isEmpty else { return }
        for track in tracks where track.lastSeenFrame == frameIndex {
            var best: Detection? = nil
            var bestOverlap: Float = 0
            for det in detections {
                if det.confidence < 0.35 { continue }
                if det.box.height < 0.10 || det.box.bottom < 0.30 { continue }
                let overlap = Self.overlap1d(track.left, track.right, det.box.left, det.box.right)
                if overlap > bestOverlap { bestOverlap = overlap; best = det }
            }
            if let best, bestOverlap >= 0.4 {
                track.vote(Labels.friendlyName(best.label))
                let truncated = best.box.top < 0.02 || best.box.bottom > 0.98
                if let known = Labels.knownHeightMeters(best.label), !truncated {
                    let boxHeightPx = best.box.height * Float(frameHeightPx)
                    track.fuseDistance(known * ground.focalLengthPx() / boxHeightPx)
                }
            }
        }
    }

    public func reset() {
        tracks.removeAll()
        frameIndex = 0
    }

    /// Intersection over the smaller extent, in 1D.
    static func overlap1d(_ l1: Float, _ r1: Float, _ l2: Float, _ r2: Float) -> Float {
        let inter = min(r1, r2) - max(l1, l2)
        if inter <= 0 { return 0 }
        let smaller = min(r1 - l1, r2 - l2)
        return smaller <= 1e-6 ? 0 : inter / smaller
    }
}

File: android/core/src/main/kotlin/com/metanav/core/ObstacleTracker.kt
package com.metanav.core

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/** A blob followed across frames so that one noisy frame can neither trigger nor cancel an alert. */
class Track internal constructor(val id: Int, blob: Blob, frameIndex: Int) {
    var left: Float = blob.left
        private set
    var right: Float = blob.right
        private set
    var distanceMeters: Float = blob.distanceMeters
        private set
    var lastSeenFrame: Int = frameIndex
        private set
    var firstSeenFrame: Int = frameIndex
        private set
    private val labelVotes = HashMap<String, Int>()

    /** Ring of frame indices in which this track was observed. */
    private val hitFrames = ArrayDeque<Int>()

    // Announcement bookkeeping, owned by the policy.
    var announcedCount: Int = 0
    var lastAnnouncedDistance: Float = Float.POSITIVE_INFINITY
    var lastAnnouncedAtMs: Long = AdvisoryPolicy.NEVER
    var lastAnnouncedUrgency: Urgency = Urgency.NONE
    var reminders: Int = 0

    val centerX: Float get() = (left + right) / 2f

    val label: String
        get() = labelVotes.maxByOrNull { it.value }?.key ?: Labels.GENERIC

    init { hitFrames.addLast(frameIndex) }

    internal fun update(blob: Blob, frameIndex: Int, window: Int) {
        left = 0.5f * left + 0.5f * blob.left
        right = 0.5f * right + 0.5f * blob.right
        distanceMeters = 0.5f * distanceMeters + 0.5f * blob.distanceMeters
        lastSeenFrame = frameIndex
        hitFrames.addLast(frameIndex)
        while (hitFrames.isNotEmpty() && hitFrames.first() <= frameIndex - window) hitFrames.removeFirst()
    }

    internal fun vote(label: String) {
        labelVotes[label] = (labelVotes[label] ?: 0) + 1
    }

    internal fun fuseDistance(meters: Float) {
        // Size-based estimates are coarse; blend gently and only when they roughly agree.
        val ratio = meters / distanceMeters
        if (ratio in 0.5f..2.0f) distanceMeters = 0.75f * distanceMeters + 0.25f * meters
    }

    fun hitsInWindow(frameIndex: Int, window: Int): Int = hitFrames.count { it > frameIndex - window }

    fun isConfirmed(frameIndex: Int, config: ReasonerConfig): Boolean =
        hitsInWindow(frameIndex, config.confirmWindow) >= config.confirmHits

    fun overlapsCorridor(config: ReasonerConfig): Boolean {
        val overlap = min(right, config.corridorRight) - max(left, config.corridorLeft)
        if (overlap <= 0f) return false
        val needed = min(0.4f * (right - left), 0.08f)
        return overlap >= needed
    }
}

class ObstacleTracker(private val config: ReasonerConfig) {
    private val tracks = ArrayList<Track>()
    private var nextId = 1
    private var frameIndex = 0

    val active: List<Track> get() = tracks

    fun currentFrame(): Int = frameIndex

    fun update(blobs: List<Blob>): List<Track> {
        frameIndex++
        val unmatched = blobs.toMutableList()
        // Greedy association by horizontal overlap, closest-first.
        for (track in tracks.sortedBy { it.distanceMeters }) {
            var best: Blob? = null
            var bestScore = 0f
            for (blob in unmatched) {
                val score = overlap1d(track.left, track.right, blob.left, blob.right)
                val centerGap = abs(track.centerX - blob.centerX)
                val s = if (score > 0.3f || centerGap < 0.15f) max(score, 0.3f - centerGap) else 0f
                if (s > bestScore) { bestScore = s; best = blob }
            }
            if (best != null) {
                track.update(best, frameIndex, config.confirmWindow)
                unmatched.remove(best)
            }
        }
        for (blob in unmatched) tracks.add(Track(nextId++, blob, frameIndex))
        tracks.removeAll { frameIndex - it.lastSeenFrame > config.maxMissedFrames }
        return tracks
    }

    /** Attach detector labels (and size-based distances) to tracks seen this frame. */
    fun applyDetections(detections: List<Detection>, ground: GroundPlaneModel, frameHeightPx: Int) {
        if (detections.isEmpty()) return
        for (track in tracks) {
            if (track.lastSeenFrame != frameIndex) continue
            var best: Detection? = null
            var bestOverlap = 0f
            for (det in detections) {
                if (det.confidence < 0.35f) continue
                if (det.box.height < 0.10f || det.box.bottom < 0.30f) continue
                val overlap = overlap1d(track.left, track.right, det.box.left, det.box.right)
                if (overlap > bestOverlap) { bestOverlap = overlap; best = det }
            }
            if (best != null && bestOverlap >= 0.4f) {
                track.vote(Labels.friendlyName(best.label))
                val known = Labels.knownHeightMeters(best.label)
                val truncated = best.box.top < 0.02f || best.box.bottom > 0.98f
                if (known != null && !truncated) {
                    val boxHeightPx = best.box.height * frameHeightPx
                    val z = known * ground.focalLengthPx() / boxHeightPx
                    track.fuseDistance(z)
                }
            }
        }
    }

    fun reset() {
        tracks.clear()
        frameIndex = 0
    }

    /** Intersection over the smaller extent, in 1D. */
    private fun overlap1d(l1: Float, r1: Float, l2: Float, r2: Float): Float {
        val inter = min(r1, r2) - max(l1, l2)
        if (inter <= 0f) return 0f
        val smaller = min(r1 - l1, r2 - l2)
        return if (smaller <= 1e-6f) 0f else inter / smaller
    }
}
